src/Parser/TypeParser.php

```diff
@@ -761,10 +798,20 @@ private function parseArrayShape(TokenIterator $tokens, Ast\Type\TypeNode $type,
 				break;
 			}
 
-			$items[] = $this->parseArrayShapeItem($tokens);
-
+			$item = $this->parseArrayShapeItem($tokens);
+			$items[] = $item;
 			$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
-		} while ($tokens->tryConsumeTokenType(Lexer::TOKEN_COMMA));
+			if (!$tokens->tryConsumeTokenType(Lexer::TOKEN_COMMA)) {
+				$done = true;
+			}
+			if ($tokens->currentTokenType() !== Lexer::TOKEN_COMMENT) {
+				continue;
+			}
+
+			$this->appendComments($item, [new Ast\Comment($tokens->currentTokenValue(), $tokens->currentTokenLine(), $tokens->currentTokenIndex())]);
```

Comment:
It's a pity that this PR adds comments only specifically to array shape items. It should be achieved somehow generally so that all types that support multiline syntax can support comments automatically.

These types come to mind:

1) Object shapes (very similar but there are different methods for them in TypeParser)
2) Conditional types
3) Union and intersection types
4) Generic types
5) Callables

Again, it should be possible to achieve this automatically without modifying parsing each of these types which can be very error-prone to implement.

I see that you added reading a comment in `enrichWithAttributes`. What happens, what kind of syntax support we lose, if we just leave the modification in enrichWithAttributes, and remove every change about parsing array shapes?

Reply:
I can try, but I think it's complicated by architectural differences between this library and `PHP-Parser`:

`PHP-Parser` has a central `getNextToken` hook in `Lexer`, and everything must go through it to get tokens. Comments are  plucked from the token list and appended to an attributes structure that is pushed onto a stack, and used to decorate nodes later.

`phpdoc-parser` requires us to manually call `enrichWithAttributes` over and over. 

Anyway, let me poke around a bit more. Maybe we can do something.

Reply:
I'm not against rearchitecting this library (in a separate PR) to make our job easier here in this PR.

Reply:
I think that could help with general comment parsing, but for the particular case of array shape items I think we're going to have to continue special-casing (because we want to pair comments specifically with items,  and not keys or values).

But if you want to start thinking about how to remove the need for all those calls to `enrichWithAttributes` it could help with general comment parsing.

Reply:
> But if you want to start thinking about how to remove the need for all those calls to enrichWithAttributes

That would be an interesting exercise! If you have ideas about this please send them in a separate PR 😊

Reply:
Well, I walked right into that one. 😆

Reply:
I'm going to try to make the same changes for object shape, but before I do that I thought it might help to make the printer test cases easier to read. In one of my other projects I made a little multi-line helper called `nowdoc`, and I added it here:

05e9eedc3c9cf58ff0a8c0cf2dbc76501222f3f1

Any objections to something like this? If so, I'll revert it, if not, I'll continue.

Reply:
Object comments seem to work:

```php
yield [
			self::nowdoc('
				/**
				 * @return object{
				 *   // your favorite bar
				 *   bar: string
				 * }
				 */'),
			self::nowdoc('
				/**
				 * @return object{
				 *   // your favorite bar
				 *   bar: string,
				 *	 // favorite foo
				 *	 foo: int
				 * }
				 */'),
			$addItemsWithCommentsToObjectShape,
		];
```

Reply:
> These types come to mind:

> Object shapes (very similar but there are different methods for them in TypeParser)
> Conditional types
> Union and intersection types
> Generic types
> Callables

Can you give an example of what each of these would look like with a comment? Don't worry, I don't intend to special-case each one, but I would like to know what you envision.

Reply:
Imagine multi-line variants of these types, and imagine there's `// something` at the end of each line.

Reply:
I'll have to experiment with Nikic parser a bit more, but I'm a little unclear if above-the-line and end-of-the-line pair a comment with the same item:

```php
array{
  // a is for apple
 a: string,
b: int
}
```

vs:

```php
array{
a: string, // a is for apple
b: int
}
```

Are these both annotating `a: string` with the same comment?

Reply:
I ran with your idea of having `TokenIterator` do more of the work, and managed to remove all the special comment handling from the array/object shape stuff, and got things working in a general sense: [0558a8b](https://github.com/phpstan/phpdoc-parser/pull/213/commits/0558a8b1341d01eb56a9c4d365f5714363aae968)

The only problem with adding comments to things like Generics and Conditionals is that they aren't written to handle multiple new lines, so I had to adjust their parsing routines to loop over new lines. 

Let me know if I'm on the right track, here, and I'll see if I can finish up.

Reply:
> Union and intersection types

I tried to get this working with comments, but multiline syntax doesn't seem to work; this doesn't work with or without comments:

```php

/**
 * @return 
 *    Foo
 *      &
 *    Bar
 * /
```

I can wok on that if you really want it, but otherwise I'd be much happier to leave it for another time.

Reply:
> Conditional types

This also won't work with comments and format-preserving printer in its current form, because it doesn't have a single `items` property of array type, so it will either need to be refactored to work more like `ArrayShapeType` and `ObjectShapeType`, or I will need to do special handling in the printer. I'm always interested in the challenge, but would love to save this for another day if you don't mind.

Reply:
> Callables

This one (nearly--had to eat extra lines) works out of the box:

[b00d20a](https://github.com/phpstan/phpdoc-parser/pull/213/commits/b00d20ac96a0ab94cc22f409c97d6519781fbde0)

Reply:
> Union

This one is nearly doable, but the code in Nikic parser seems to forbid it:

```php
// We go multiline if the original code was multiline,
                // or if it's an array item with a comment above it.
                if ($insertStr === ', ' &&
                    ($this->isMultiline($origNodes) || $arrItem->getComments())
                ) {
                    $insertStr = ',';
                    $insertNewline = true;
                }
```

So, nikic parser only prints comments when it's a newline, and we only decide to do a new line when there are comments, or the insert string is `,`, but for union the insert string is `|`. I can tweak the logic if you want, but I think it will make us diverge from nikic parser.

At this point, I think I have done just about everything I can. Please advise.

Reply:
> Are these both annotating a: string with the same comment?

Finally, I did more debugging with Nikic parser, and no, it seems that comments that come after an array item are _not_ paired with the item that precedes it; they are associated with the following line:

```php
$items = [
    'a' => 'foo',  // puppies are cute
    'b' => 'bar'
];
```

In this example, the "puppies" comment is bound to the `'b' => 'bar'` line.